lib/oplog/tail.go

```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
 
 		var rawData bson.Raw
 
-		for iter.Next(&rawData) {
-			ts, pubs := tailer.unmarshalEntry(rawData)
+		for {
+			gotResult, didTimeout, didLosePosition, err := readNextFromCursor(query)
 
-			if ts != nil {
-				lastTimestamp = *ts
-			}
+			if gotResult {
+				decodeErr := query.Decode(&rawData)
+				if decodeErr != nil {
+					log.Log.Errorw("Error decoding oplog entry", "error", decodeErr)
+
+				}
+
+				ts, pubs := tailer.unmarshalEntry(rawData)
+
+				if ts != nil {
+					lastTimestamp = *ts
+				}
+
+				for _, pub := range pubs {
+					if pub != nil {
+						out <- pub
+					} else {
+						log.Log.Error("Nil Redis publication")
+					}
+				}
+			} else if didTimeout {
+				log.Log.Info("Oplog cursor timed out, will retry")
 
-			for _, pub := range pubs {
-				if pub != nil {
-					out <- pub
-				} else {
-					log.Log.Error("Nil Redis publication")
+				query, queryErr = issueOplogFindQuery(oplogCollection, lastTimestamp)
+
+				if queryErr != nil {
+					log.Log.Errorw("Error issuing tail query", "error", queryErr)
+					return
 				}
+
+				break
+			} else if didLosePosition {
+				// Our cursor expired. Make a new cursor to pick up from where we
+				// left off.
+				query, queryErr = issueOplogFindQuery(oplogCollection, lastTimestamp)
+
+				if queryErr != nil {
+					log.Log.Errorw("Error issuing tail query", "error", queryErr)
+					return
+				}
+
+				break
+			} else if err != nil {
```

Comment:
It's up to you, but rather than a longer if/else if ..../else block, something like the `else if err != nil` check could go right after line 176, and handling errors like that in an early-return block seems more go-ish